crates/ra_ide/src/diagnostics.rs

```diff
@@ -123,14 +123,72 @@ pub(crate) fn diagnostics(db: &RootDatabase, file_id: FileId) -> Vec<Diagnostic>
             severity: Severity::Error,
             fix: Some(fix),
         })
+    })
+    .on::<hir::diagnostics::NoSuchField, _>(|d| {
+        res.borrow_mut().push(Diagnostic {
+            range: sema.diagnostics_range(d).range,
+            message: d.message(),
+            severity: Severity::Error,
+            fix: missing_struct_field_fix(&sema, file_id, d),
+        })
     });
+
     if let Some(m) = sema.to_module_def(file_id) {
         m.diagnostics(db, &mut sink);
     };
     drop(sink);
     res.into_inner()
 }
 
+fn missing_struct_field_fix(
+    sema: &Semantics<RootDatabase>,
+    file_id: FileId,
+    d: &hir::diagnostics::NoSuchField,
+) -> Option<Fix> {
+    let record_expr = sema.ast(&d.field, d.file);
+
+    let record_lit = ast::RecordLit::cast(record_expr.syntax().parent()?.parent()?)?;
+    let def = sema.resolve_path(&record_lit.path()?)?;
```

Comment:
hmm it would probably be more accurate to use the variant resolution from `infer`... maybe we need to add a `resolve_variant` to `Semantics`

Reply:
Does `resolve_path` not correctly resolve enum variants?
To make sure: you're talking about making [this `resolve_variant` function](https://github.com/rust-analyzer/rust-analyzer/blob/36353bb1827dbd2efcde2d18c8598c4cc5e2e296/crates/ra_hir_ty/src/infer.rs#L419) accessible through `Semantics`?

Reply:
> Does resolve_path not correctly resolve enum variants?

I think it will at least not resolve things like `Self { ... }`, `Self::Variant { ... }` or `T::SomeAssociatedType { ... }` (that last one is still a FIXME though), probably some more edge cases.

I don't mean exposing that `resolve_variant`; rather, the `InferenceResult` contains that information in `variant_resolutions`, so the `resolve_variant` on `Semantics` will work similarly to `resolve_field` or `resolve_method_call`.

Reply:
Is [this](https://github.com/rust-analyzer/rust-analyzer/pull/4851/files#diff-7b6f4630b83a99b8e5a8473750af7c7eR257) what you were thinking of?

Reply:
Yes!